### PR TITLE
New mongo package options to optimize Oplog tailing performance to include/exclude certain collections

### PR DESCRIPTION
**Pull request by Twisterking**
Hello all,

Following several conversations on the drawbacks of Meteor's oplog implementation and its limitations when it comes to scaling (see: [my forums post](https://forums.meteor.com/t/cpu-spikes-due-to-oplog-updates-without-subscriptions/60028), [a slightly older blog post](https://blog.meteor.com/tuning-meteor-mongo-livedata-for-scalability-13fe9deb8908) and a [2 year old github discussion](https://github.com/meteor/meteor/discussions/11842)), I investigated in the core a bit and propose an easy solution which, I hope, should work for many users who have issues with this.

Please consider our situation at Orderlion: We have several collections, where huge amounts of data are being updated on a daily basis. On all of these collections we have not a single publication, for these we don't care about any reactivity.

Due to Meteor's Oplog implementation, even in these cases a big batch insert/update of data results often in very big sustained CPU spikes on the meteor app server.

Internal tests I did with my suggested implementation have shown, **that the CPU load due to Oplog tailing collections with big batch updates, can be reduced about 100x with just completely ignoring those collections**. 
During an import of about 600k documents in a test collection, I could reduce the CPU load from an average 25-30% down to approx. 0,2 - 0,3%.

The usage is actually very straight forward: Consider a `settings.json` file like this:

```js
{
  // ... all other settings,
  "public": {
    // all your public settings
  },
  "packages": {
    "mongo": {
      "oplogExcludeCollections": ["products", "prices"] // This would exclude both collections "products" and "prices" from any oplog tailing.
      // OR:
      "oplogIncludeCollections": ["messages"] // ONLY watch the "messages" collection for oplog updates (and all "admin.$cmd" oplogs)
    }
  }
}
```

Please feel free to comment but I really hope this can be merged soon. It would help us at Orderlion a lot and solve many unwanted server restarts (server is auto-restarting because it becomes unresponsive due to the sustained high CPU load).

Cheers, Patrick

### CODE REVIEW

File: packages/mongo/oplog_tailing.js
Comment (radekmie) [inline packages/mongo/oplog_tailing.js:254]: IMO it should be called `oplogIncludeCollections`.

File: packages/mongo/oplog_tailing.js
Comment (radekmie) [inline packages/mongo/oplog_tailing.js:273]: Why not this instead:
```suggestion
      oplogSelector.ns = {
         $regex: oplogSelector.ns,
         $in: excludeCollections.map((collName) => `${self._dbName}.${collName}`)
       }
```

_(Of course the two branches will be almost identical then, just asking for further simplification.)_

**Review by radekmie — COMMENTED** [review]
I think the most important part will be to add tests. At least:
1. Configure both and see if it fails.
2. Configure inclusion and check if non-included ones are not processed.
3. Configure exclusion and check if non-excluded ones are processed.

**Comment by harryadel** [comment]
Why can the configuration be set from settings.json and not the code? 
```js
// This is pseudo code so don't judge
import { mongo } from 'meteor/mongo'

mongo.oplogExcludeCollections = [];
mongo.oplogWatchCollections = [];

```
 I've noticed this trend in Meteor lately and I believe it's getting out of hand. The settings file is getting bloated with config. It's supposed to hold ENV variables/secrets. :man_shrugging: 
 
 cc @zodern

**Comment by Twisterking** [comment]
@harryadel 
In my opinion the settings file is the right place for this. The oplog tailing is setup very very early during the startup of your app. Finding the right place to put your suggested code so that your settings are defined "on time" is not straight forward in my opinion. 

The settings file is kinda "always loaded". 

Also I tried to follow the logic that already exists for the mongo package - it already uses the Meteor.settings for other package options.

**Comment by harryadel** [comment]
@Twisterking I understand why you went this path. I'm merely discussing the idea not trying to put you down :)

>Finding the right place to put your suggested code so that your settings are defined "on time" is not straight forward in my opinion.

But isn't that what `Meteor.startup` is for?

>Also I tried to follow the logic that already exists for the mongo package - it already uses the Meteor.settings for other package options.

Yeah, this trend has been going on for a while so naturally you continued down this trail. I just don't think this path  is the best IMO.

**Comment by Twisterking** [comment]
@harryadel understood, no worries! No harm done!
I am aware of Meteor.startup of course, but I was not sure if this works in this case because the mongo package indeed is very much core. 
Anyways ...

I will write some tests as suggested and will also create a PR for the meteor docs so this new setting is documented. 

Any other feedback? As I said I would love to get this PR ready for merging asap! :+1

File: packages/mongo/oplog_tailing.js
Comment (Twisterking) [reply packages/mongo/oplog_tailing.js:273]: I am sorry I dont fully follow here. With your suggestion, this is basically an $and. 
But this wont work for the "include variant", as you need to support both admin.$cmd logs AND the user's included collections - so you actually need an $or in the end. 
What am I missing?

Comment (radekmie) [reply packages/mongo/oplog_tailing.js:273]: You're right 🤦 Sure, let's leave it as it is.

**Comment by radekmie** [comment]
@Twisterking I'm for leaving it in the settings. Doing it in the code would be problematic, as it'd need to execute before the connection is established and that's _deep_.

@harryadel The `Meteor.startup` is for executing code _after_ the server starts, so it's too late. Even top-level in the app is too late (you can execute database operations there already).

**Comment by Twisterking** [comment]
> I think the most important part will be to add tests. 

@radekmie I am very happy to do that! The issue is, I don't really have any experience with `tinytest` yet.
I will make that work somehow, but: **What I do not know (and where I will need your support)** is, how can I define a `Meteor.settings` entry when running the Oplog tests? I tried to just define it somehow like this:

```js
process.env.MONGO_OPLOG_URL && Tinytest.addAsync(
  'mongo-livedata - oplog - options.oplogExcludeCollections',
  async test => {
    if (!Meteor.settings.packages) Meteor.settings.packages = {};
    if (!Meteor.settings.packages.mongo) Meteor.settings.packages.mongo = {};
    Meteor.settings.packages.mongo.oplogExcludeCollections = ['foobar'];
    console.log('oplogExcludeCollections', Meteor.settings.packages.mongo);
    test.equal(true, true);
  }
);
```

This of course works, but it is "too late" - the oplog file itself is already loaded and there, `Meteor.settings.packages.mongo` is undefined! 😢 

--> **How do I define a Meteor setting in a test "early enough"?**

**Comment by radekmie** [comment]
The only easy option I see would be to change the `Meteor.settings`, create a new instance of `OplogHandle`, test it, and revert the `Meteor.settings`.

**Comment by Twisterking** [comment]
I am sorry, but I am not familiar enough with the meteor core to do this based on this comment! :/ I will need some more detailed help please! Maybe @zodern can help?

File: packages/mongo/oplog_tests.js
Comment (Twisterking) [inline packages/mongo/oplog_tests.js:193]: @radekmie might need your help here to make this work please! :)

File: packages/mongo/oplog_tests.js
Comment (Twisterking) [inline packages/mongo/oplog_tests.js:201]: Without using my custom `OplogHandle`, this console.log works just fine! So there has to be something wrong with my implementation of setting `myOplogHandle` to be used with the `defaultRemoteCollectionDriver`.

**Comment by zodern** [comment]
This will be very helpful for some apps. I'm not very familiar with this code, but I have a couple questions.

1. Meteor tracks the last oplog entry it's seen to know where to resume tailing the oplog. If there's been a large number of newer oplog entries for databases or collections that are excluded, could resume tailing cause performance issues if mongo has to filter out the unneeded docs? This happens every time Meteor does not receive an oplog doc for 30 seconds, there is a failover, or in some other situations.

To give an example in case my description isn't clear. Let's say the database that is being observed is updated every 10 minutes on average, and there is another database for metrics that has hundreds of updates per second. Whenever Meteor restarts tailing the oplog (which would happen almost every 30 seconds since there are rarely oplog entries for the observed database), it would query for all of the relevant docs newer than the last oplog entry it received. Since the last oplog entry Meteor received might have been up to 10 minutes ago, there could be a large number of other oplog entries mongo would have to filter.

2. Would it be possible to make this automatic, at least for the databases? Meteor could track which databases have had observers created for them, and update the query whenever a new database is observed. It seems like as long as Meteor tracks the timestamp of the last oplog entry it processed, there shouldn't be any downside to restarting tailing with a new query. Meteor already does it frequently with the same query.

This would probably involve OplogHandle tracking the last ts it's seen: https://github.com/meteor/meteor/blob/c26569d224eb081b9aef950f44066c555ca9b6f9/packages/mongo/oplog_tailing.js#L256
and then stopping and recreating the oplog tail handle as needed. I'm not sure if there would be a performance issue with listing too many databases or collections.

3. What happens if someone creates an observer on an excluded database or collection? Will the observer not provide updates after the initial find, or will it fallback to polling?

**Comment by Twisterking** [comment]
@zodern 
Thanks for your reply! I indeed also think that this change is actually quite simple, but can have a huge impact in some cases like ours. It **really** changes the performance/scaling of Meteor for us at Orderlion significantly!

ad 1) good point indeed, to be honest I don't know. 
ad 2) same here. Maybe I can ask you of any other maintainer to help here a bit? :)
ad 3) I thought about this too. Also, don't know yet. I am happy to test it! In my opinion both is fine, as long as these 2 new options are properly documented. It is really meant for collections where there are no subscriptions whatsoever. But it makes sense that maybe polling of course is better than nothing.

Is there anyone else who is willing to contribute here? I might be biased of course but I honestly believe this change can help with some Meteor scaling issues a lot for some projects out there!

**Comment by Twisterking** [comment]
Okay, so I finally managed to make the tests work! 🎉 
Also, I implemented a `console.warn` whenever a subscription on an excluded or not included collection is created.

Please have a look!

**Comment by zodern** [comment]
> ad 1) good point indeed, to be honest I don't know.
ad 2) same here. Maybe I can ask you of any other maintainer to help here a bit? :)
ad 3) I thought about this too. Also, don't know yet. I am happy to test it! In my opinion both is fine, as long as these 2 new options are properly documented. It is really meant for collections where there are no subscriptions whatsoever. But it makes sense that maybe polling of course is better than nothing.

For 1), it might be worth mentioning in the docs that it would be inefficient if the watched collections are rarely updated, and other collections have a very large number of updates.

2) I might create a package to experiment with this and some other ideas

3) There are quite a few scenarios where Meteor is unable to use the oplog for an observer, and Meteor always falls back to polling in those situations. For consistency, it would be good if Meteor did the same thing here. Throwing an error or logging a warning might also work, though it should be done when creating the observer instead of in `_publishCursor` to handle all of the ways an observer can be created. The relevant function for either solution (falling back to polling, or throwing an error) would probably be https://github.com/meteor/meteor/blob/c26569d224eb081b9aef950f44066c555ca9b6f9/packages/mongo/mongo_driver.js#L1352

**Comment by rj-david** [comment]
On the discussion on where to place the configuration, in terms of developer experience, is it possible to do this in the definition of the collection (also where you normally add the schema). So when a developer creates a new collection, the developer can configure that collection for oplog tailing or not on the same file. Moving it outside the creation of the collection will result for most developers to forget to disable/enable oplog tailing.

Just an idea and of course if it is even possible

**Comment by Twisterking** [comment]
@zodern: Thank you, I will have a look into `_observeChanges` today!
@rj-david: to be honest I think this is too late in the lifecycle! I thought about this approach too, but whenever you create collections, the Oplog handle/cursor has already been created. So this approach won't work afaik. That's why, again, I went for the `Meteor.settings` approach in the end.

**Comment by Twisterking** [comment]
@zodern @radekmie 
Does anyone of the Meteor team have infos, when this PR might be merged and released? 😃 
We are really desperately waiting for its release! ✌️

**Comment by harryadel** [comment]
While I'm not part of the core team but if @zodern @radekmie can give the green light since they're responsible for reviewing, @denihs @Grubba27 can move in.

File: docs/source/api/collections.md
Comment (radekmie) [inline docs/source/api/collections.md:1091]: 1. The wording here is unnecessarily pessimistic. Ideally @Grubba27 could rephrase it.
2. I'd rather go with "operations", not "transactions", as the latter may suggest actual database transactions (which is irrelevant here). It could be also generalized to "write operations" too.

File: packages/mongo/oplog_tests.js
Comment (radekmie) [inline packages/mongo/oplog_tests.js:194]: It'd be good to clear those settings after the test not to influence others.

File: packages/mongo/oplog_tests.js
Comment (radekmie) [inline packages/mongo/oplog_tests.js:197]: Same goes for the `OplogHandle`.

File: packages/mongo/oplog_tests.js
Comment (radekmie) [inline packages/mongo/oplog_tests.js:279]: Typo?

File: packages/mongo/oplog_tests.js
Comment (radekmie) [inline packages/mongo/oplog_tests.js:284]: As I wrote above - this should be handled in `oplogOptionsTest`.

File: packages/mongo/package.js
Comment (radekmie) [inline packages/mongo/package.js:105]: It should not be needed if the `oplogOptionsTest` would clear after itself.

**Review by radekmie — CHANGES_REQUESTED** [review]

**Comment by radekmie** [comment]
@zodern 
> 1. Meteor tracks the last oplog entry it's seen to know where to resume tailing the oplog. If there's been a large number of newer oplog entries for databases or collections that are excluded, could resume tailing cause performance issues if mongo has to filter out the unneeded docs? This happens every time Meteor does not receive an oplog doc for 30 seconds, there is a failover, or in some other situations.

It may be a visibly long operation, but it should not be an issue - oplog collection is usually already in memory, so it'll be really fast to grep through.

> 2. Would it be possible to make this automatic, at least for the databases? Meteor could track which databases have had observers created for them, and update the query whenever a new database is observed. It seems like as long as Meteor tracks the timestamp of the last oplog entry it processed, there shouldn't be any downside to restarting tailing with a new query. Meteor already does it frequently with the same query.

It could, sure. I'd rather make it in a separate step to get this PR out sooner.

File: packages/mongo/oplog_tests.js
Comment (Twisterking) [reply packages/mongo/oplog_tests.js:194]: I am actually doing just that?! Please just look at the very bottom of `oplogOptionsTest()` :)

File: packages/mongo/oplog_tests.js
Comment (Twisterking) [reply packages/mongo/oplog_tests.js:197]: I can only reset the used `OplogHandle` like so, but I don't think there is any way to "destroy" `myOplogHandle`, is there?!

Reset:
```js
MongoInternals.defaultRemoteCollectionDriver().mongo._setOplogHandle(defaultOplogHandle);
```

File: packages/mongo/oplog_tests.js
Comment (Twisterking) [reply packages/mongo/oplog_tests.js:284]: different in this case, because for this test `oplogOptionsTest()` actually throws an error because of using both options - like it should! That's why here the teardown actually happens outside of `oplogOptionsTest()`!

File: packages/mongo/package.js
Comment (Twisterking) [reply packages/mongo/package.js:105]: But it does not hurt either! ;)

**Comment by Twisterking** [comment]
@radekmie thanks for your feedback! PR updated!

File: packages/mongo/oplog_tests.js
Comment (radekmie) [reply packages/mongo/oplog_tests.js:194]: No, you _delete_ it - I suggest _resetting_ it to what it was. I know that right now it will be the same, but it may change in the future.

File: packages/mongo/oplog_tests.js
Comment (radekmie) [reply packages/mongo/oplog_tests.js:197]: My point is that you should extract the previous handle before setting the new one and set the old one back after the test.

File: packages/mongo/oplog_tests.js
Comment (radekmie) [reply packages/mongo/oplog_tests.js]: That only suggests that the cleanup should happen in a `finally` block in `oplogOptionsTest` - then it'll happen even if the test fails.

File: packages/mongo/package.js
Comment (radekmie) [reply packages/mongo/package.js:105]: That's a matter of personal preference - I suggest not to touch anything not related to the PR to make it easier to review and then easier to compare (or bisect if needed).

File: packages/mongo/oplog_tests.js
Comment (Twisterking) [reply packages/mongo/oplog_tests.js:197]: That is exactly what I am doing!

Comment (radekmie) [reply packages/mongo/oplog_tests.js:197]: No, you're extracting it on the top level, not in the test. That's a significant difference, as the top-level code executes first only to register the tests.

_(Also, I don't know how to understand your exclamation marks... I hope you're excited and not getting angry at me.)_

File: packages/mongo/package.js
Comment (Twisterking) [reply packages/mongo/package.js:105]: okay, I will move it again!

Comment (Twisterking) [reply packages/mongo/package.js]: Fixed!

File: packages/mongo/oplog_tests.js
Comment (Twisterking) [reply packages/mongo/oplog_tests.js]: Also fixed!

**Comment by Twisterking** [comment]
@radekmie one more update done! Thanks again for your feedback!

File: packages/mongo/oplog_tests.js
Comment (radekmie) [inline packages/mongo/oplog_tests.js:231]: Nope, we don't 😛 
```suggestion
```

File: packages/mongo/collection_tests.js
Comment (radekmie) [inline packages/mongo/collection_tests.js:213]: Why are those suddenly needed?

File: packages/mongo/oplog_tests.js
Comment (Twisterking) [reply packages/mongo/oplog_tests.js:231]: Then you please have to help me! Otherwise I can't make the last test work, where I want to test that it fails if both settings are used at the same time. If I catch the error here, then I can't "listen" for an error in the actual test!

Comment (radekmie) [reply packages/mongo/oplog_tests.js:231]: I meant you can just `try { ... } finally { ... }`. Is there a problem with that?

Comment (Twisterking) [reply packages/mongo/oplog_tests.js]: Allllright, my bad 😁 you are right, `try { ... } finally { ... }` works just fine! ✌️

**Review by radekmie — APPROVED** [review]

**Comment by Twisterking** [comment]
Hello everyone! Thanks for approving the PR @radekmie !
Any news on a possible release date? As stated, this is really blocking us more and more! :(

**Comment by StorytellerCZ** [comment]
I would be for Meteor 2.16 with https://github.com/meteor/meteor/pull/12436 asap.

**Comment by Twisterking** [comment]
> > Hello everyone! Thanks for approving the PR @radekmie ! Any news on a possible release date? As stated, this is really blocking us more and more! :(
> 
> You can already use this without waiting for a new release since the changes are in the package and not in the core

How? The version of the package was not released yet, no? Sorry, don't fully understand how this works! Also, I forget to increase the version number of the `mongo` package! my bad!

What exactly do I need to do to use this version of the `mongo` core package?

**Comment by rj-david** [comment]
> What exactly do I need to do to use this version of the `mongo` core package?

https://guide.meteor.com/writing-atmosphere-packages#local-packages

**Comment by Twisterking** [comment]
> > What exactly do I need to do to use this version of the `mongo` core package?
> 
> [guide.meteor.com/writing-atmosphere-packages#local-packages](https://guide.meteor.com/writing-atmosphere-packages#local-packages)

That won't work here! I tried to install my package as a local package! The issue is, that all the other packages still used the "default mongo package" behind the scenes, and not my "custom" local package. Couldn't get it to work when I tried it.

**Comment by rj-david** [comment]
Short of releasing this package, the things that can be tested are:

1. Incrementing the version
2. Also forking those packages that used mongo locally

**Comment by StorytellerCZ** [comment]
If you use meteor from local checkout with this branch that should do the trick if local package override is failing.

**Comment by Twisterking** [comment]
Any updates on when the release of 2.16 will happen? Hopefully incl. this PR please! ✌️

**Comment by Twisterking** [comment]
Updates? Is there a planned release date for 2.16 yet? We really need this change asap ...

**Comment by nachocodoner** [comment]
> Updates? Is there a planned release date for 2.16 yet? We really need this change asap ...

No estimated date yet. First, we're focused on the 3.x RC. Meanwhile, we aim to merge candidates and make quick gains into the 2.16 branch. The PR has been updated [to provide more details about the plan](https://github.com/meteor/meteor/pull/13066#issue-2200350442). We'll announce a date as soon as we have one.

**Comment by jeffreyflynt** [comment]
> > > What exactly do I need to do to use this version of the `mongo` core package?
> > 
> > 
> > [guide.meteor.com/writing-atmosphere-packages#local-packages](https://guide.meteor.com/writing-atmosphere-packages#local-packages)
> 
> That won't work here! I tried to install my package as a local package! The issue is, that all the other packages still used the "default mongo package" behind the scenes, and not my "custom" local package. Couldn't get it to work when I tried it.

For future reference, I ran into this issue, what I had to do was name my local package a, so it's loaded before everything else.

For instance, I have this configured for Mongo:

```
const secureContext = tls.createSecureContext({
    ca: TLS_CA,
    cert: TLS_CRT,
    key: TLS_KEY,
    minVersion: "TLSv1.3",
  });

  process.env.MONGO_URL = MONGO_URL;

  Mongo.setConnectionOptions({
    tls: true,
    secureContext,
    compressors: ["zstd"],
    family: 4,
  });

  ```

**Comment by Colkadome** [comment]
I have a strange behaviour after the update if anyone knows whats going on here.

Pretend I need to update a user property in the `Accounts.onLogin`, and I also have a collection "UserStats" that I update, but I also have included UserStats in the `oplogExcludeCollections` setting (as I don't have any publications that use this collection):

settings.json:
```
{
  "packages": {
    "mongo": {
      "oplogExcludeCollections": [
        "UserStats"
      ]
    }
  }
}
```

The login method takes around ~15 seconds to respond in the websocket connection with this setting on, AND with the following code in the onLogin callback:

```
Accounts.onLogin((data) => {
  Meteor.users.update(...);
  UserStats.update(...);
});
```

Either removing "UserStats" from `oplogExcludeCollections` will fix this, OR by changing the update code to use async calls (both need to be async). I can also include some driver refresh call and it will be fine:

```
Accounts.onLogin(async (data) => {
  
  await Meteor.users.rawCollection().updateOne(...).catch(...);
  Meteor.users._driver.mongo._refresh(Meteor.users._name, { ... });

  await UserStats.rawCollection().updateOne(...).catch(...);
  UserStats._driver.mongo._refresh(UserStats._name, { ... });

});
```

I have solved the issue now. I don't have much time at the moment to test more conditions here, but I thought I would just comment this in case anyone else is experiencing something similar. There may be some steps in the update() calls that I am ignoring with the above code but for my use-case it is fine.

**Pull request merged**
